### PR TITLE
Added scripts in dynamic_tests_v2

### PR DESCRIPTION
**Pull request by AshburnLee**
### 添加脚本
1. Add masked_select.py for **masked_select OP**
- 根据'class PytorchAPIBenchmarkBase'中函数'Variable'，当x_dtype=int时，requires_grad=Fasle，但是pytorch表示需要计算grad，所以如果config中的x_dtype为int时，转换成float。
- 当输入类型为整型时，pytorch不做反向计算，为了对齐，此种情况，paddle也不做反向计算。
2. Add pad3d.py for **pad OP**
- torch.nn.functional.pad()中的'reflect' mode 未实现，而torch.nn.ReplicationPad3d()的reflection 模式逻辑与paddle中reflect 模式逻辑不同，此mode时API不对齐。
- 分别调用PaddlePad2d()和TorchPad2d()，并且对于输入做了检查，保证是5维输入。
3. Add expand.py for **expand_v2 OP**
4. Add expand_as.py for **expand_as_v2 OP**

### 修复脚本错误

5. 原来脚本test2_v/lstm.py，报错segmentation fault。使用nn.LSTM API后，脚本正确执行。

### CODE REVIEW

File: api/dynamic_tests_v2/lstm.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+class TorchLstm(PytorchAPIBenchmarkBase):
+    def build_graph(self, config):
+        dtype_map = {
+            "float16": torch.float16,
+            "float32": torch.float32,
+            "float64": torch.float64,
+            "int32": torch.int32,
+            "int64": torch.int64,
+            "bool": torch.bool
+        }
```
Comment (GaoWei8) [inline api/dynamic_tests_v2/lstm.py:87]: 参数转换可以在def to_pytorch中进行定义

File: api/dynamic_tests_v2/rnn.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+class TorchRnn(PytorchAPIBenchmarkBase):
+    def build_graph(self, config):
+        dtype_map = {
+            "float16": torch.float16,
+            "float32": torch.float32,
+            "float64": torch.float64,
+            "int32": torch.int32,
+            "int64": torch.int64,
+            "bool": torch.bool
+        }
```
Comment (GaoWei8) [inline api/dynamic_tests_v2/rnn.py:53]: 上同

File: api/tests_v2/configs/lstm.json [hunk 3]
```diff
@@ -32,5 +32,6 @@
             "type": "bool",
             "value": "False"
         }
-    }
+    },
+    "atol": 1
```
Comment (GaoWei8) [inline api/tests_v2/configs/lstm.json:36]: 为什么阈值需要设置为1

File: api/tests_v2/configs/rnn.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        "num_layers": {
+            "type": "int",
+            "value": "2"
+        },
+        "time_major": {
+            "type": "bool",
+            "value": "False"
+        }
+    },
+    "atol": 1.5
```
Comment (GaoWei8) [inline api/tests_v2/configs/rnn.json:27]: 阈值设置的过大

File: api/tests_v2/lstm.py [hunk 5]
```diff
@@ -44,20 +50,17 @@ def build_program(self, config):
...
-            init_c=init_c,
-            max_len=config.max_len,
+        rnn = paddle.nn.LSTM(
+            input_size=config.input_shape[-1],
             hidden_size=config.hidden_size,
             num_layers=config.num_layers,
-            dropout_prob=0.2,
-            is_bidirec=config.is_bidirec)
+            dropout=0.0,
+            direction=direct)
```
Comment (GaoWei8) [inline api/tests_v2/lstm.py:58]: paddle.fluid.layers.lstm接口已经被废弃了，既然改成新接口，原有的lstm.json文件也可以尝试修改下。

File: api/dynamic_tests_v2/avg_pool2d.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+class PDAvgPool2d(PaddleDynamicAPIBenchmarkBase):
+    def build_graph(self, config):
+        x = self.variable(name='x', shape=config.x_shape, dtype=config.x_dtype)
+        result = paddle.nn.functional.avg_pool2d(
+            x=x,
+            kernel_size=config.kernel_size,
+            stride=config.stride,
+            padding=config.padding,
+            ceil_mode=config.ceil_mode,
+            data_format=config.data_format)
```
Comment (GaoWei8) [inline api/dynamic_tests_v2/avg_pool2d.py:27]: 2.0的api，max_pool2d和avg_pool2d可能整合成一个api：pool2d。可以确认下。

File: api/dynamic_tests_v2/lstm.py (hunk 1 above)
Comment (AshburnLee) [reply api/dynamic_tests_v2/lstm.py:87]: DONE.

File: api/dynamic_tests_v2/rnn.py (hunk 2 above)
Comment (AshburnLee) [reply api/dynamic_tests_v2/rnn.py:53]: DONE.

File: api/tests_v2/lstm.py (hunk 5 above)
Comment (AshburnLee) [reply api/tests_v2/lstm.py:58]: DONE.

File: api/dynamic_tests_v2/rnn.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+class TorchRnn(PytorchAPIBenchmarkBase):
+    def build_graph(self, config):
+        config.input_shape = [
+            config.input_shape[1], config.input_shape[0], config.input_shape[2]
+        ]
+        input = self.variable(
+            name="input", shape=config.input_shape, dtype=config.input_dtype)
+        tensor_h = torch.empty(config.init_h_shape, dtype=config.init_h_dtype)
+        init_h = torch.nn.init.constant_(tensor=tensor_h, val=0.0)
```
Comment (Xreki) [inline api/dynamic_tests_v2/rnn.py:65]: 可使用torch.full或torch.zeros

**Comment by Xreki** [comment]
> Add remainder.py for elementwise_mod OP
Note：对于pytorch，input和other的精度如果是浮点型，系统的requires_grad为True，但是pytorch该API的反向计算，似乎没
有实现。所以json文件中的x，y的类型为设int。

elementwise_mod并没有注册反向op，测试脚本中也没有调用append_gradients，所以没有必要修改json配置。如果是因为requires_grad设置，那测试代码里面改成requires_grad=False即可。

File: api/dynamic_tests_v2/pad3d.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        # So set run_torch as False when using 'reflect' mode
+        if self.mode == "reflect":
+            print(
+                "Warning:\n"
+                "  1. 'reflect' mode in torch.nn.functional.pad is not implemented.\n"
+            )
+            self.run_torch = False
+
+
+class PaddlePad3d(PaddleDynamicAPIBenchmarkBase):
```
[the remark this answers is not in the document]
Comment (AshburnLee) [reply api/dynamic_tests_v2/pad3d.py:37]: Done.

File: api/dynamic_tests_v2/pad3d.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class Pad3dConfig(APIConfig):
+    def __init__(self):
+        super(Pad3dConfig, self).__init__('pad3d')
+
+    def init_from_json(self, filename, config_id=0, unknown_dim=16):
+        super(Pad3dConfig, self).init_from_json(filename, config_id,
+                                                unknown_dim)
+        # in API 'torch.nn.functional.pad()', 'reflect' mode is not implemented,
```
[the remark this answers is not in the document]
Comment (AshburnLee) [reply api/dynamic_tests_v2/pad3d.py:25]: Done.

File: api/dynamic_tests_v2/masked_select.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        super(MaskedSelectConfig, self).__init__("masked_select")
+
+    def init_from_json(self, filename, config_id=0, unknown_dim=16):
+        super(MaskedSelectConfig, self).init_from_json(filename, config_id,
+                                                       unknown_dim)
+        # according to 'class PytorchAPIBenchmarkBase', in fucntion 'Variable'
+        # requires_grad=False, when x_dtype==int. However, this API requires 
+        # GRAD, so we do the following cast:
+        if self.x_dtype == "int32":
+            self.x_dtype = "float32"
```
[the remark this answers is not in the document]
Comment (AshburnLee) [reply api/dynamic_tests_v2/masked_select.py:29]: pytorch返回："only Tensors of floating point and complex dtype can require gradients"，所以向将int改为float或删去config.backward

File: api/dynamic_tests_v2/masked_select.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        mask = self.variable(
+            name="mask", shape=config.mask_shape, dtype=config.mask_dtype)
+        result = torch.masked_select(input=x, mask=mask)
+
+        self.feed_list = [x, mask]
+        self.fetch_list = [result]
+        # only Tensors of floating point and complex dtype can require gradients
+        if config.x_dtype in ["float16", "float32", "float64"]:
+            if config.backward:
+                self.append_gradients(result, [x])
```
Comment (Xreki) [inline api/dynamic_tests_v2/masked_select.py:44]: 这种写法不赞成，因为使用collect_api_info.py收集API信息的时候，一个API是否有backward，只能是True或者False。不支持一些配置是True，一些配置是False。支持这种case，可以通过设置`variable(..., stop_gradient=True)`。

File: api/tests_v2/configs/remainder.json [hunk 12]
```diff
@@ -0,0 +1,43 @@
+[{
```
Comment (Xreki) [inline api/tests_v2/configs/remainder.json:1]: 删掉这个文件。

**Review by Xreki — APPROVED** [review]
有2个小的修改意见，下个PR再改吧。
